open-metadata-implementation/governance-servers/open-lineage-services/open-lineage-services-server/src/main/java/org/odpi/openmetadata/governanceservers/openlineage/services/GraphServices.java

```diff
@@ -96,45 +102,78 @@ private String glossary(Graph graph, String guid) {
     }
 
     /**
-     * Returns a subgraph containing all paths leading from any root node to the queried node. The queried node can be a
-     * column, table, or host.
+     * Returns a subgraph containing all paths leading from any root node to the queried node, but condensed so that
+     * the path itself between the root nodes and the queried node has been abstracted.
      *
-     * @param scope The scope queried by the user: hostview, tableview, columnview.
+     * @param scope The scope queried by the user: tableview, columnview.
      * @param graph MAIN, BUFFER, MOCK, HISTORY.
-     * @param guid  The guid of the node of which the lineage is queried of. This can be a column, table, or host node.
+     * @param guid  The guid of the node of which the lineage is queried of. This can be a column or a table.
      * @return a subgraph in the GraphSON format.
      */
     private String ultimateSource(String scope, Graph graph, String guid) {
         GraphTraversalSource g = graph.traversal();
         String edgeLabel = getEdgeLabel(scope);
 
-        Graph subGraph = (Graph)
-                g.V().has(GraphConstants.PROPERTY_KEY_ENTITY_GUID, guid).
-                        until(inE(edgeLabel).count().is(0)).
-                        repeat(inE(edgeLabel).subgraph("subGraph").outV()).
-                        cap("subGraph").next();
-        return janusGraphToGraphson(subGraph);
+        List<Vertex> sourcesList = g.V().has(GraphConstants.PROPERTY_KEY_ENTITY_GUID, guid).
+                until(inE(edgeLabel).count().is(0)).
+                repeat(inE(edgeLabel).outV()).dedup().toList();
+        Vertex originalQueriedVertex = g.V().has(GraphConstants.PROPERTY_KEY_ENTITY_GUID, guid).next();
+
+        Graph responseGraph = TinkerGraph.open();
```

Comment:
In this method a new temp graph is created where a subset of nodes/properties/edges are copied. This could be a performance issue as the memory and processor consumption could get high, creating and populating a graph is not trivial. We should investigate for a better way as we only need a subset of nodes and edges, with no manipulation operations on them. This could be done in future PR, this one can be merged as it is from my side.